### ISSUE TITLE
FSR 2 has pixel fizzling when in fullscreen mode

### ISSUE THREAD

**Issue by Jamsers** [godotengine/godot#83191]
### Godot version

v4.2.dev6.official [57a6813bb]

### System information

Godot v4.2.dev6 - Windows 10.0.22621 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3060 Laptop GPU (NVIDIA; 31.0.15.3758) - AMD Ryzen 5 5600H with Radeon Graphics (12 Threads)

### Issue description

When in Fullscreen or Exclusive Fullscreen mode, FSR 2 exhibits extra fizzling that is not present when in any of the Windowed modes.

Fullscreen:
![fullscreen1](https://github.com/godotengine/godot/assets/39361911/d7d71e9e-c193-4fc3-a9cf-2739f6a9bcca)
Windowed:
![windowed1](https://github.com/godotengine/godot/assets/39361911/99d0cfe6-e39f-4721-96ee-08704f125400)
Fullscreen:
![fullscreen2](https://github.com/godotengine/godot/assets/39361911/4cf3b9f5-d5ef-4427-b35c-ba9ac2a8ba65)
Windowed:
![windowed2](https://github.com/godotengine/godot/assets/39361911/a8b21278-c2f0-42b5-b6d7-3c18ba80cd78)
Fullscreen:
![fullscreen3](https://github.com/godotengine/godot/assets/39361911/8adbd4be-c245-4243-aa9f-2af28cfcba4f)
Windowed:
![windowed3](https://github.com/godotengine/godot/assets/39361911/002697d2-7aef-4fc7-8899-e2566bfe91b7)

Apologies for low quality images, had to take these from a video recording since it's hard to capture through screenshots.

### Steps to reproduce

Launch the [repro project](https://github.com/Jamsers/Bistro-Demo-Tweaked/tree/FSR2Fizzle) and set the render resolution to half of the native resolution. (Not required, but it makes the issue easier to spot)

Push the scooter around in a shadowed area.
Keep the character in sunlight and run in front of a shadowed wall.
Enable noclip mode and fly around on one of the flat, bright rooftops.

Do these again while setting the project Window Mode to Windowed or Maximized, and you'll notice the fizzling is gone.

### Minimal reproduction project

Test project is too big to upload, but issue can be reproduced in this branch of the project's Github repo.
[Bistro-Demo-Tweaked/FSR2Fizzle](https://github.com/Jamsers/Bistro-Demo-Tweaked/tree/FSR2Fizzle)

**Comment by akien-mga** [comment]
CC @DarioSamo

**Comment by DarioSamo** [comment]
I don't see the possible relation between windowed mode or fullscreen? Are you sure this isn't just different resolutions? There's like no differences between code paths accounting for that.

**Comment by Jamsers** [comment]
>I don't see the possible relation between windowed mode or fullscreen?

Me neither! I had to test this a lot and record gameplay and scrub through footage to make sure I wasn't imagining it 😅

What's weird is that you'd think it'd happen in exclusive fullscreen mode, but not in borderless windowed mode, since borderless windowed mode should be similar to normal windowed mode, just without the borders.

But nope! Happens in borderless windowed mode as well. What an odd bug. I wonder if it happens on Mac or Linux as well. I'll do some more Windows specific testing on my end, maybe it doesn't happen on an AMD GPU for example.

**Comment by DarioSamo** [comment]
> > I don't see the possible relation between windowed mode or fullscreen?
> 
> Me neither! I had to test this a lot and record gameplay and scrub through footage to make sure I wasn't imagining it 😅
> 
> What's weird is that you'd think it'd happen in exclusive fullscreen mode, but not in borderless windowed mode, since borderless windowed mode should be similar to normal windowed mode, just without the borders.
> 
> But nope! Happens in borderless windowed mode as well. What an odd bug. I wonder if it happens on Mac or Linux as well. I'll do some more Windows specific testing on my end, maybe it doesn't happen on an AMD GPU for example.

I've noticed your project disables V-Sync. Are you getting significantly different framerates in both modes?

I've spotted some usual artifacts and a trail when moving the scooter in a particular way, but nothing unlike what happens with FSR2 usually. But I also get around over 200 FPS while on fullscreen, so that does compensate a lot for some of the issues.

**Comment by DarioSamo** [comment]
I tested all the scenarios you mentioned and notice no fizzling. I'm on an RTX 3090 Ti on Windows 11 FWIW with a 1080p 60 Hz monitor. Tested both exclusive fullscreen and the windowed modes and I don't see any difference (except borderless fullscreen performs absolutely awful for whatever reason, seems we need some investigation there).

It's worth pointing out the FSR2 implementation isn't much more than just hooking the relevant inputs to what the library requests, so some issues might be inherent to how the algorithm works. I think what might be different here is either your target resolution or your framerate rather than the Window mode itself, as it's not causing any issues on my end by itself.

**Comment by Jamsers** [comment]
Some further testing I've done:
Tested on an AMD GPU (the integrated GPU on my laptop), noticed the same problem.
Tested on a separate PC, noticed the same problem.

Something interesting I noted when testing on my iGPU, the problem is more noticeable the higher the refresh rate is.

Potential differentiating factor: In PresentMon, both Windowed and Borderless Windowed modes present as `Composed: Flip`, only Exclusive Fullscreen mode presents as `Hardware: Independent Flip`. (on the AMD iGPU it presents as `Hardware: Legacy Flip`)

I can't see how that could possibly affect FSR 2 though so don't know what to make of that

>I've noticed your project disables V-Sync. Are you getting significantly different framerates in both modes?

Similar framerates, both around 100fps on my system.

>I've spotted some usual artifacts and a trail when moving the scooter in a particular way

The trail may be this bug - I initially had the same impression as you that it's standard FSR 2/TAA artifacting but what made me consider it a bug is that the trailing doesn't happen in Windowed mode. Did it look like bright pixel fireflies against the darker background?

>except borderless fullscreen performs absolutely awful for whatever reason, seems we need some investigation there

Probably not related to this bug but I found this interesting, on my system the performance was completely similar on both modes, the only difference was slightly higher input lag on the windowed modes for me

>sounds like a problem with the particular target resolution rather than the window mode itself. Maybe we need to match that number to reproduce it.

In my testing I used 540p in both Fullscreen and Windowed mode.
![image](https://github.com/godotengine/godot/assets/39361911/0d9d1cef-ec90-4201-abd8-9b3ac57dc036)
It should render (internally) at that target resolution regardless of window size or native monitor resolution, the way I've implemented it.

**Comment by DarioSamo** [comment]
> The trail may be this bug - I initially had the same impression as you that it's standard FSR 2/TAA artifacting but what made me consider it a bug is that the trailing doesn't happen in Windowed mode. Did it look like bright pixel fireflies against the darker background?
Yeah, most of the artifacts usually present themselves on high contrast scenarios. This is an inherent artifact to how FSR2 works based on a luminance-based mask for locking pixels according to their page. Not something I actually had to implement myself, but it's something I've noticed across Godot, other projects where I've implemented it and even commercial games. It can even show up with DLSS in areas with low brightness as well.

> > sounds like a problem with the particular target resolution rather than the window mode itself. Maybe we need to match that number to reproduce it.
> In my testing I used 540p in both Fullscreen and Windowed mode. ![image](https://user-images.githubusercontent.com/39361911/274603394-0d9d1cef-ec90-4201-abd8-9b3ac57dc036.png) It should render (internally) at that target resolution regardless of window size or native monitor resolution, the way I've implemented it.

Ah, I was sticking to the values presented by the UI. I wonder if we're running into a situation where not using the recommended multipliers can cause potential issues. Seems like it'll warrant further investigation in your project to see if it's actually picking that resolution (will confirm with RenderDoc later).

**Comment by Jamsers** [comment]
>I wonder if we're running into a situation where not using the recommended multipliers can cause potential issues

Just wanna note through that in my testing it happens at all resolutions (even native) - I just lowered it to 540p because the fatter pixels show it better

>There might be a possibility the internal renderer is not picking the resolution you want

Yeah maybe some floating point weirdness, since I can't actually directly set the internal resolution I get to my desired resolution with some math and setting `viewport_set_scaling_3d_scale()`

The relevant code is found in `/Scripts/UIController.gd`, relevant functions are `_on_viewport_resize()` and `get_resolution_actual()`

```gdscript
# example resolution target
var renderTargetVertical = 540.0

func _on_viewport_resize():
	RenderingServer.viewport_set_scaling_3d_scale(get_viewport().get_viewport_rid(), renderTargetVertical/float(get_resolution_actual()))

func get_resolution_actual():
	var yToUse
	var project_width = float(ProjectSettings.get_setting("display/window/size/viewport_width"))
	var project_height = float(ProjectSettings.get_setting("display/window/size/viewport_height"))
	var viewport_width = float(get_viewport().size.x)
	var viewport_height = float(get_viewport().size.y)
	
	# the project Stretch Aspect mode is set to keep_height
	# check if window aspect ratio is taller than project aspect ratio
	# if it is, 3D image will be letterboxed so figure out the true 3D vertical res
	
	if viewport_width/viewport_height < project_width/project_height:
		yToUse = (viewport_width/project_width) * project_height
	else:
		yToUse = viewport_height
		
	return int(yToUse)
```

**Comment by Jamsers** [comment]
I've managed to isolate the most problematic model (the scooter) and created a minimal reproduction project to showcase the issue in isolation. I've edited the original issue with an upload of the repro project and videos showcasing the issue in it.

**Comment by DarioSamo** [comment]
I suspect the scooter's problems might be related to the metallic material. Have you tried messing with that and see if it's the reflections being so bright that cause the issue to appear?

**Comment by Jamsers** [comment]
>I suspect the scooter's problems might be related to the metallic material

Almost certainly, but I just chose the scooter because it showcased the issue the best.

The bug here honestly isn't the fizzling artifact itself but rather the different behavior between Fullscreen and Windowed mode.

**Comment by DarioSamo** [comment]
Fair enough, seems like we should determine what exactly is different.

Considering the renderer creates the textures on its own and doesn't rely on the swap chain texture other than for the final step, I think I'd suspect the resolution itself ending up in some uneven number being problematic (e.g. 1279 vs 1280 pixels). This should be easily verifiable with something like RenderDoc. I'll attempt that when I get some more time.

**Comment by mrjustaguy** [comment]
Try increasing the similar_threshold value in https://github.com/godotengine/godot/blob/master/thirdparty/amd-fsr2/shaders/ffx_fsr2_lock.h to see if the issue persists. I've found in my test scenes that going from the default 1.05 to 1.5 eliminated all the super ghosts

**Comment by Jamsers** [comment]
>Try increasing the similar_threshold value in https://github.com/godotengine/godot/blob/master/thirdparty/amd-fsr2/shaders/ffx_fsr2_lock.h

Just tested it, didn't make a difference in Bistro-Demo-Tweaked unfortunately 😕

**Comment by mrjustaguy** [comment]
well 1.5 worked for my scenes, try setting the value really, really high to effectively disable it.. also make sure that it recompiles it! for some reason it often just ignores changes to the file.. I've had to work around it by causing an error in https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shaders/effects/fsr2/fsr2_lock_pass.glsl by renaming the file to a non-existant one (by adding an a at the end) and after that fails, reverting the change and compiling again.

**Comment by Jamsers** [comment]
>also make sure that it recompiles it! for some reason it often just ignores changes to the file

I did `scons --clean` and then built it, you were right, it does change it! On Bistro-Demo-Tweaked, at `similar_threshold = 1.5f`, the pixel fizzling is *significantly* reduced... however, it's not completely eliminated - there's still tiny fizzling that happens. Also, the fizzling still completely goes away when you switch from fullscreen mode to windowed.

**Comment by mrjustaguy** [comment]
you can increase the value up to 10 see if it completely eliminates it, my scenes don't have highly contrasting speculars and are mostly diffuse.

and the explanation for the fullscreen-windowed is the slight difference in pixel positions that may change this. I've tested at different resolutions in windowed and got vastly different results. so this by itself means nothing.. you would also see different results at various different window resolutions so..

**Comment by mrjustaguy** [comment]
hey, an even better test, set https://github.com/godotengine/godot/blob/13a0d6e9b253654f5cc2a44f3d0b3cae10440443/thirdparty/amd-fsr2/shaders/ffx_fsr2_lock.h#L102

to false, and do a clean compile. this will effectively kill locking completely as it'll always return false, and no new locks will be created. I've done so and now even in my tests in High Contrast scenes there's no more problems

**Comment by Jamsers** [comment]
Interesting, but what does locking do anyway? Surely there's some kind of tradeoff or change happening when you set the threshold high or turn it off completely, otherwise why even have the code in the first place?

**Comment by mrjustaguy** [comment]
Without locking, the upscaling part is borked.

**Comment by Jamsers** [comment]
>Without locking, the upscaling part is borked

well now, best not then 😅

>Sadly it falls on AMD to resolve this by improving the Locking mechanism

Here's the thing though, I've never seen this artifact in any other game or engine that's implemented FSR 2. If you know of any please tell but I've genuinely only seen this in Godot and I wouldn't be surprised if this only really happens in Godot.

Disocclusion artifacts, transparency jittering, ghosting, you could definitely argue these are issues with FSR 2 itself because these artifacts happen in other games and engines with FSR 2 as well. But not this one.

If multiple games and engines use a library but an issue only shows on one of them, it seems reasonable to deduce that the issue is either with engine itself, or its implementation/usage of the library, right?

**Comment by mrjustaguy** [comment]
Not really, as this is heavily depends on content, and also most games only expose it as an upscaling option, which provides artifacts of it's own that helps hide this issue..

I also have seen this in other games, Immortals of Avium "specular shimmering" is actually I believe bad Pixel locking in action in a UE5 title.

**Comment by Jamsers** [comment]
>most games only expose it as an upscaling option

This issue also shows up in Godot even when upscaling. In fact a curious fact about this artifacting is that the fizzling fireflies seem to render at native res even when upscaling from a sub-native resolution.

>I also have seen this in other games, Immortals of Avium "specular shimmering" is actually I believe bad Pixel locking in action

This is the same thing @DarioSamo said but you guys are mistaken, I just didn't have the energy to argue because he seemed dead set on using it as a reason not to tackle this bug so I shifted attention to the discrepancy between fullscreen and windowed instead.

This artifact is different from the ones you guys are describing - the artifacts you guys are describing are just run of the mill disocclusion or motion vector artifacts inherent in TAA. The Immortals of Aveum artifacts you point out, for example, are caused by the fact that the weapon loses motion vectors while it's animating. (likely due to it being animated with a shader rather than a skeleton)

They present similar to this bug, but this bug specifically is about the bright and distinct pixel size fireflies that fizzle and remain for a time when bright highlights pass over dark areas. The ghosting that also tends to follow in its wake, I agree that's not a bug, that's just TAA being TAA.

**Comment by mrjustaguy** [comment]
1) Wrong video / video part, not talking about weapon, but about the chains

2) The fireflies ARE the pixel locking in action, and not your run of the mill TAA issue

3) Ofc this shows up even when upscaling, it just makes the issue more difficult to spot due to other issues not present when running at native.

**Comment by Jamsers** [comment]
>Wrong video / video part, not talking about weapon, but about the chains

Can you give me a timestamp? I can't seem to find a part with chains showing the issue

**Comment by mrjustaguy** [comment]
must find the video first, dont have the time right now, working on testing FSR 2 fixes i'm trying to make to the locking mechanism

**Comment by mrjustaguy** [comment]
OK so, got a different result by changing the division thing, but hard to tell if it's better..

For now I'm just going to be disabling the pixel locking entirely for my builds as I'm only using it as AA, not as an upscaling solution so the pixel locking doesn't really add much to the table aside from the issue.

**Comment by EzraT** [comment]
I can confirm this issue.

Info: Godot v4.2.2.stable - Fedora Linux 40 (KDE Plasma) - Wayland - Vulkan (Forward+) - dedicated AMD Radeon RX 580 Series (RADV POLARIS10) - Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz (8 Threads)

Whether its in fullscreen or not doesn't seem to matter in my case, happens either way.
I also tested my project on a Steam Deck in the same manner, and it doesn't happen there, so this might be a hardware specific problem.

Video:

https://github.com/godotengine/godot/assets/781121/d2bc5748-9d5a-4fc4-b809-fc6f1a200871

**Comment by mrjustaguy** [comment]
try current master, Reverse-Z could have changed FSR's results (don't try dev 6 due to FSR MVs being broken, that was patched after that)

**Comment by EzraT** [comment]
> try current master, Reverse-Z could have changed FSR's results (don't try dev 6 due to FSR MVs being broken, that was patched after that)
> 
> Also, this looks like the artifacts FSR 3.1 will resolve - see https://gpuopen.com/events/amd-at-gdc-2024/ If that is the case, we'll have to wait for AMD to finally release it to the public before it can get integrated into Godot, which they said will happen in Q2, so anywhere between now and the next month and 2 weeks, if they don't delay it.

Indeed, it's pretty much unnoticable in my project in the current master. Good to see this is being improved.
The OP reproduction project seems the same for me though.

**Comment by mrjustaguy** [comment]
well as hypothesized before, seems like very tiny differences in the math cause this issue to appear, and it seems likely that the FSR 3.1 patch will resolve it, especially now that i think about it, the demo they've shown is actually showing the same artifact, just present during upscaling instead of native which in my experience does reduce the impact of this issue, but that is likely because other issues start to appear.

**Comment by mrjustaguy** [comment]
Implementing https://github.com/godotengine/godot-proposals/issues/13787 solves this issue by adding an option to opt out of the problematic bit of the Algorithm. Note This is only a Viable option for Native, as Upscaling Suffers a bit, whereas native only gets a tad bit more shimmer but otherwise looks fine.
Also Soft Shadows don't leave any trailing with the Locking Disabled, unlike the current state that requires Ultra Soft Shadows to minimize soft shadow issues for FSR 2

This is an issue that is also present with 3.1 but the algorithm has changed and the results may be a bit different, and opting out of the mechanism will be more difficult for it, but it also handles some cases better and others worse (like soft shadows can be used at lower settings better compared to FSR 2, but light retention may hit larger surface areas)

There's also some indication that FSR 4 still suffers from this issue to a lesser extent based on my limited Star Wars Outlaws testing on my Brother's RX 9060 XT 16gb, so this is a deeply rooted issue for AMD's Upscaling/Antialiasing Technology.

I have no reference point for DLAA, and proper XMX XeSS, but dp4a XeSS and TSR have both had some similar tendencies observed in Everspace 2 at native, but nowhere near to the same extent, so it seems like an issue inherent to Upscaling Technologies in general, especially considering the Native TAA doesn't show any such tendencies
